### PR TITLE
Build R documentation using Docker

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

Build the R documentation using docker to make it easier to run and test + simplify the CircleCI workflow.

## How is this patch tested?

- Confirm the R documentation can be built without errors.
- Run our internal Jenkins job for documentation build against this PR's branch and confirm it succeeds.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/R/mlflow/Dockerfile.dev [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+WORKDIR /mlflow/mlflow/R/mlflow
+RUN apt-get update -y
+RUN apt-get install git wget libxml2-dev libgit2-dev -y
+RUN TEMP_DEB=$(mktemp) && \
+    wget --directory-prefix $TEMP_DEB https://github.com/jgm/pandoc/releases/download/2.16.2/pandoc-2.16.2-1-amd64.deb && \
+    dpkg --install $(find $TEMP_DEB -name '*.deb') && \
+    rm -rf $TEMP_DEB
+COPY DESCRIPTION .
+COPY .install-deps.R .
+RUN Rscript -e 'source(".install-deps.R", echo = TRUE)'
```
Comment (harupy) [inline mlflow/R/mlflow/Dockerfile.dev:12]: Renamed `Docker.build` to `Docker.dev` and add packages required to build the docs.

File: docs/source/R-api.rst [hunk 2]
```diff
@@ -1309,13 +1309,16 @@ to be used by package authors to extend the supported MLflow models.
...
++-----------+---------------------------------------------------------+
+| Argument  | Description                                             |
++===========+=========================================================+
+| ``model`` | The loaded MLflow model flavor.                         |
++-----------+---------------------------------------------------------+
+| ``data``  | A data frame to perform scoring.                        |
++-----------+---------------------------------------------------------+
+| ``...``   | Optional additional arguments passed to underlying      |
+|           | predict methods.                                        |
++-----------+---------------------------------------------------------+
```
Comment (harupy) [inline docs/source/R-api.rst:1321]: I updated the version of pandoc to `2.16.2` from `2.7.1`, which added some changes in this file.

File: .circleci/config.yml [hunk 3]
```diff
@@ -2,56 +2,11 @@ version: 2.1
...
-      - run:
-          name: Dump R dependencies
-          working_directory: mlflow/R/mlflow
-          command: |
-            Rscript .dump-r-dependencies.R
-
-      - restore_cache:
-          keys:
-            - r-cache-{{ checksum "mlflow/R/mlflow/R-version" }}
-
```
Comment (harupy) [inline .circleci/config.yml:54]: Removed these steps because the docker image contains all required tools.

File: .circleci/config.yml [hunk 4]
```diff
@@ -2,56 +2,11 @@ version: 2.1
 
 jobs:
   build_doc_r:
-    docker:
-      - image: ubuntu:20.04
+    machine:
+      image: ubuntu-2004:202111-01
```
Comment (harupy) [inline .circleci/config.yml:6]: To use docker volume mounting, we need to use the machine executor:
https://support.circleci.com/hc/en-us/articles/360007324514-How-can-I-use-Docker-volume-mounting-on-CircleCI-

File: mlflow/R/mlflow/.build-doc.R [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Install Rd2md from source as a temporary fix for the rendering of code examples, until
+# a release is published including the fixes in https://github.com/quantsch/Rd2md/issues/1
+# Note that this commit is equivalent to commit 6b48255 of Rd2md master
+# (https://github.com/quantsch/Rd2md/tree/6b4825579a2df8a22898316d93729384f92a756b)
+# with a single extra commit to fix rendering of \link tags between methods in R documentation.
+devtools::install_git("https://github.com/smurching/Rd2md", ref = "mlflow-patches")
```
Comment (WeichenXu123) [inline mlflow/R/mlflow/.build-doc.R:6]: Shall we use commit hash instead of branch which may be changed ?

File: mlflow/R/mlflow/Dockerfile.dev [hunk 6]
```diff
@@ -0,0 +1,12 @@
+FROM rocker/r-ver:4.1.2
+
+WORKDIR /mlflow/mlflow/R/mlflow
+RUN apt-get update -y
+RUN apt-get install git wget libxml2-dev libgit2-dev -y
+RUN TEMP_DEB=$(mktemp) && \
+    wget --directory-prefix $TEMP_DEB https://github.com/jgm/pandoc/releases/download/2.16.2/pandoc-2.16.2-1-amd64.deb && \
+    dpkg --install $(find $TEMP_DEB -name '*.deb') && \
```
Comment (WeichenXu123) [inline mlflow/R/mlflow/Dockerfile.dev:8]: Can we install pandoc by `RUN apt-get install pandoc -y` ?

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: mlflow/R/mlflow/.build-doc.R (hunk 5 above)
Comment (harupy) [reply mlflow/R/mlflow/.build-doc.R:6]: Good point, but we usually don't update this branch.

File: mlflow/R/mlflow/Dockerfile.dev (hunk 6 above)
Comment (harupy) [reply mlflow/R/mlflow/Dockerfile.dev:8]: The version installed by pandoc is 2.5 which is too old and contains a bug.

Comment (harupy) [reply mlflow/R/mlflow/Dockerfile.dev:9]: I inserted a comment.

**Pull request merged**